pkg/dockerregistry/server/writelimiter/semaphore.go

```diff
@@ -0,0 +1,24 @@
+package writelimiter
+
+type semaphore struct {
+	c chan struct{}
+}
+
+func newSemaphore(count int) *semaphore {
+	return &semaphore{
+		c: make(chan struct{}, count),
+	}
+}
+
+func (s *semaphore) Up() {
+	<-s.c
+}
+
+func (s *semaphore) TryDown() bool {
```

Comment:
Names here should be consistent with other rate limiters - either "TryIncrement" and "Decrement" or something that is consistent.

However, since this is a prime candidate for leaks, it would be better for this to be structured in a way that cannot fail.

```
func (s *maxRateLimit) Try(fn func()) {
  take
  defer release
  fn()
}
```

But I think the behavior we want is to actually take and wait up to a max, not immediately exit.  Injecting limited delay is acceptable.